### PR TITLE
Added intersect type simulator

### PR DESCRIPTION
**Pull request by mark-todd**
I had an idea for how we can draft intersection behaviour - I've been testing it and so far it works well! It's based on option 5 from the Any discussion. The idea is that this simulates the attributes of an intersection (using runtime). If you request an attribute from the resulting intersection, it returns the *type* of the requested attribute. I'm thinking that this will let people make tests easily and see what type their attribute would be in the intersection. By agreeing the behaviour of this simulator, we in turn agree the behaviour of the type checker. Here's a quick example:

```py
from typing import Any

from intersection_examples import Intersection


class A:
    test: int

    def foo(self) -> int:
        pass


class B:
    test: int

    def foo(self) -> int:
        return 1


test = Intersection[A, B]
print(test) # B & A
print(test.test)# <class 'int'>
print(test.foo)# (self) -> int


test_new = Intersection[A, Any]
print(test_new) # Any & A
print(test_new.foo) # (self) -> int
print(test_new.x) # typing.Any
```
Note that the comments here are the actual printed results in runtime of the simulator.

### CODE REVIEW

**Comment by mark-todd** [comment]
The simulator now supports all cases I've thought of so far, including combinations with existing overloads.

**Comment by tomasr8** [comment]
I don't have much time to have a look right now but this looks really cool! I wonder how difficult it would to take this and create a fork of mypy that actually implements intersections (option 5 or other). I'm sure we'll discover more edge cases that need to be ironed out.

**Comment by tomasr8** [comment]
Yeah the mypy codebase is pretty inscrutable :sweat_smile: I'll see if I have the time over Christmas to have a look

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -11,11 +11,6 @@ repos:
     hooks:
     -   id: black
 
--   repo: https://github.com/pycqa/isort
-    rev: 5.12.0
-    hooks:
-      - id: isort
-        name: isort (python)
```
Comment (CarliJoy) [inline .pre-commit-config.yaml:18]: Why did you remove isort?

Comment (mark-todd) [reply .pre-commit-config.yaml:18]: Ah yeah I couldn't get it to pass the pre-commit check on the line:
```py
from inspect import Signature, signature
```
I think isort might not be case-sensitive because it kept swapping these - I tried other solutions like the below:
```py
from inspect import Signature
from inspect import signature as sig_func
```
but that still didn't work - I just concluded having isort wasn't worth the trouble

Comment (mark-todd) [reply .pre-commit-config.yaml:18]: I might be that I've missed something though here - or perhaps that isort needs to use the `black` profile

Comment (CarliJoy) [reply .pre-commit-config.yaml:18]: The whole point of `isort` is to sort and not for you to manually sort.

Not production repo, not important.

In general, thanks for the work you put into this. Look promising but honestly I don't get what you doing, too many dunder methods. Some comments / docstrings would help me to get a bit more of an idea.
Also document what you want to implement in the code.

Is think you implantation covers currently only how an attribute is read `Intersection[T,S].x` but not how it is set, i.e. `Intersection[T, S].x = int #  valid or not?` can't be determined, right?

Comment (mark-todd) [reply .pre-commit-config.yaml:18]: > In general, thanks for the work you put into this. Look promising but honestly I don't get what you doing, too many dunder methods. Some comments / docstrings would help me to get a bit more of an idea.
Also document what you want to implement in the code.

Sure, I can add some docs to explain it - I appreciate there's quite a bit of magic method craziness in there!

> Is think you implantation covers currently only how an attribute is read Intersection[T,S].x but not how it is set, i.e. Intersection[T, S].x = int #  valid or not? can't be determined, right?

Yeah this is only intended to be used with getattribute - I could add a setattribute override, but I'm not sure what it would mean in the context of the simulator. The idea is that when you get an attribute it gets you the attribute *type* rather than the attribute itself. I can add a nice error message though for if anyone tries to set an attribute
